desktop/components/json_page/client/editor.coffee

```diff
@@ -90,4 +94,60 @@ module.exports = class JSONPageEditor
         attachRemove $el = $(this).prev().clone()
         $(this).before $el
 
+    @addSortingArrows()
     this
+
+  #
+  # Adds the ability to sort array items with arrows.
+  # Unfortunately Hulk.js doesn't provide this functionality or the ability to
+  # re-render based on data manipulations so we have to jump through hoops to
+  # associate the DOM to `@data` and update the ordering for both.
+  # TODO: We could probably avoid this classic problem with a
+  # React based solution: https://github.com/arqex/react-json
+  #
+  addSortingArrows: ->
+    $('.hulk-array-element').prepend "
+      <div class='json-page-array-header'>
+        <div class='json-page-array-header-up'>▲</div>
+        <div class='json-page-array-header-down'>▼</div>
+      </div>
+    "
+    $('.json-page-array-header-up').click @moveArrayItem('up')
+    $('.json-page-array-header-down').click @moveArrayItem('down')
+
+  moveArrayItem: (dir) => (e) =>
+    # Find the array's key and index this arrow is a part of in @data
+    arrayKey = $(e.target)
+      .closest('.hulk-array')
+      .siblings('.hulk-map-key')
+      .val()
+    index = $(e.target).closest('.hulk-array-element').index()
+    targetIndex = index + if dir is 'up' then -1 else 1
+
+    # Traverse to the top of the DOM tree, storing the `keys` needed
+    # to then traverse down @data and grab the associated data reference
+    keys = [arrayKey]
+    goUpALevelFrom = ($el) =>
+      $map = $el.closest('.hulk-map')
+      $parentVal = $map.closest('.hulk-map-value-container')
+      if $parentVal.length
+        keys.unshift $parentVal.siblings('.hulk-map-key').val()
+        goUpALevelFrom $map
+      else
+        array = keys.reduce ((acc, part) => acc && acc[part]), @data
+
+        # Now that we've found the reference in @data swap the two items
+        # in @data and re-render keeping the list expanded
+        [item, target] = [array[index], array[targetIndex]]
+        array[index] = target
+        array[targetIndex] = item
+        @render()
+        $(".hulk-map-key[value=#{arrayKey}]")
+          .siblings('.hulk-collapse-item')
+          .removeClass('collapsed')
+          .html('Collapse')
+          .siblings('.hulk-array')
+          .show()
+        $('.hulk-preview-iframe').get().contentDocument.location.reload(true)
```

Comment:
Hoops indeed! But this works for me, it's clear what you're doing here and I appreciate how recursion made this easier to understand. 

Side note, I just listened to this podcast about Steve Ballmer, the "Developers developers developers" guy, and he's all about hoops now so we're good: 
![image](https://cloud.githubusercontent.com/assets/2236794/26470689/98deffa6-416c-11e7-95fc-eba40b6aaf3c.png)